ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables.hs

```diff
@@ -0,0 +1,388 @@
+instance HasLedgerTables (Ticked1 l) => HasTickedLedgerTables l
+
+-- | LedgerTables are projections of data from a LedgerState and as such they
+-- can be injected back into a LedgerState. This is necessary because the Ledger
+-- rules are unaware of UTxO-HD changes. Thus, by stowing the ledger tables, we are
+-- able to provide a Ledger State with a restricted UTxO set that is enough to
+-- execute the Ledger rules.
+--
+-- In particular, HardForkBlock LedgerStates are never given diretly to the
+-- ledger but rather unwrapped and then it is the inner ledger state the one we
+-- give to the ledger. This means that all the single era blocks must be an
+-- instance of this class, but HardForkBlocks might avoid doing so.
+type CanStowLedgerTables :: LedgerStateKind -> Constraint
+class CanStowLedgerTables l where
+
+  stowLedgerTables     :: l ValuesMK -> l EmptyMK
+  default stowLedgerTables ::
+       (LedgerTablesAreTrivial l)
+    => l ValuesMK
+    -> l EmptyMK
+  stowLedgerTables = convertMapKind
+
+  unstowLedgerTables   :: l EmptyMK  -> l ValuesMK
+  default unstowLedgerTables ::
+       (LedgerTablesAreTrivial l)
+    => l EmptyMK
+    -> l ValuesMK
+  unstowLedgerTables = convertMapKind
+
+{-------------------------------------------------------------------------------
+  Serialization Codecs
+-------------------------------------------------------------------------------}
+
+-- | This class provides a 'CodecMK' that can be used to encode/decode keys and
+-- values on @'LedgerTables' l mk@
+--
+-- TODO: can this be removed in favour of EncodeDisk and DecodeDisk?
+type CanSerializeLedgerTables :: LedgerStateKind -> Constraint
+class CanSerializeLedgerTables l where
+  codecLedgerTables :: LedgerTables l CodecMK
+  default codecLedgerTables ::
+       ( FromCBOR (Key l), FromCBOR (Value l)
+       , ToCBOR   (Key l), ToCBOR   (Value l)
+       )
+    => LedgerTables l CodecMK
+  codecLedgerTables = LedgerTables $ CodecMK toCBOR toCBOR fromCBOR fromCBOR
+
+-- | Default encoder of @'LedgerTables' l ''ValuesMK'@ to be used by the
+-- in-memory backing store.
+valuesMKEncoder ::
+     ( HasLedgerTables l
+     , CanSerializeLedgerTables l
+     )
+  => LedgerTables l ValuesMK
+  -> CBOR.Encoding
+valuesMKEncoder tables =
+       CBOR.encodeListLen (ltcollapse $ ltmap (K2 . const 1) tables)
+    <> ltcollapse (ltliftA2 (K2 .: go) codecLedgerTables tables)
+  where
+    go :: CodecMK k v -> ValuesMK k v -> CBOR.Encoding
+    go (CodecMK encK encV _decK _decV) (ValuesMK m) =
+         CBOR.encodeMapLen (fromIntegral $ Map.size m)
+      <> Map.foldMapWithKey (\k v -> encK k <> encV v) m
+
+-- | Default decoder of @'LedgerTables' l ''ValuesMK'@ to be used by the
+-- in-memory backing store.
+valuesMKDecoder ::
+     ( HasLedgerTables l
+     , CanSerializeLedgerTables l
+     )
+  => CBOR.Decoder s (LedgerTables l ValuesMK)
+valuesMKDecoder = do
+    numTables <- CBOR.decodeListLen
+    if numTables == 0
```

Comment:
Indeed, I don't think the `if` is necessary, but not because `lttraverse` would evaluate to `ltpure emptyMK`. Currently, it will always be the case that

```haskell
ltcollapse (ltmap (K2 . const 1) ret) == 1
```

So if `numTables == 0` then the assertion would check `1 == 0`

The proper thing to do is probably to remove the `if` statement, and to replace the `Exn.assert` by a proper decoder failure using `fail`

Reply:
And as suggested before, we should use a fold instead of a collapse

Reply:
https://github.com/IntersectMBO/ouroboros-consensus/pull/1267#discussion_r1856433804